feat: Reworked ItemCompareFeature

uTa4u opened the pull request:
Changes inside of ItemCompareFeature class:
- Automatically compares with equipped items (including accessories and held weapon)
- Only compare with items of the same type
- Item selection indicator is correctly rendered, unlike in old version
- More than one item can be selected. Selected items can be of different types - feature will only pick items of the same type
- Tooltips don't clip outside of screen and among themselves even on gui scale 3
- Removes lore and set info from items during comparison to safe vertical space. Both can be toggled off. Set stats are not affected
- Displays tags above items for ease of distinction between equipped/selected/hovered items. Can be toggled off. Russian lang file included
- Config option to enable centered item names. Since I'm sure players will install LegendaryTooltips and then report lack of centered item names during comparison as a bug

Other changes:
- New method in ItemUtils to check if item is `empty accessory slot`
- Refactored InventoryModel to use aforementioned method
- New helper methods in TooltipUtils
- A few patterns made public in WynnItemParser
-  ItemStatInfoFeature and TooltipFittingFeature now use methods from TooltipUtils

uTa4u committed:
feat: Reworked ItemCompareFeature

uTa4u committed:
ci: spotless formatting

magicus commented:
Do you have a screenshot or screencast that showcases the functionality in this PR?

uTa4u commented:
> Do you have a screenshot or screencast that showcases the functionality in this PR?

Sure, here

https://github.com/user-attachments/assets/9063500d-5e1e-41a5-894f-b95025480621

magicus reviewed (approved):
I think this looks overall good, but I'd like for @kristofbolyai to take a look as well.

kristofbolyai commented on common/src/main/java/com/wynntils/features/tooltips/ItemCompareFeature.java:
```diff
@@ -4,162 +4,590 @@
-                .max(Integer::compareTo)
-                .orElse(0);
+        List<Component> hoveredLines = new ArrayList<>(event.getTooltips());
+        if (removeFlavourText.get()) removeFlavourText(hoveredLines);
```
We add brackets for every if that is multi-line, or does not have a flow-control statement (return/continue/break) as it's expression. Please go through this method and fix them.

kristofbolyai commented on common/src/main/java/com/wynntils/features/tooltips/ItemCompareFeature.java:
```diff
@@ -4,162 +4,590 @@
+    private static final String HOVERED_SELECTED_KEY = "feature.wynntils.itemCompare.tag.hovered_selected";
+
+    // First equippedCount items in itemsToCompare will have "Equipped" tag, others will have "Selected" tag
+    private int equippedCount = 0;
```
Essentially this could be changed so these items are kept track of in a separate list from Selected items. I think that would make it less error-prone, as it'd be much harder to "miss-count" items accidentally. The current implementation is fair, and is likely to not miss-count, but in the future, I think this code would be just a tiny bit more maintainable if we didn't have to keep track of the count in this way.

uTa4u replied:
If you don't mind, I will leave it as it is for now. I'm planning to implement comparing tomes and charms in the future and do some refactoring at the same time.

kristofbolyai replied:
Okay, I can let this slip, it's not a huge issue IMO.

kristofbolyai commented on common/src/main/java/com/wynntils/features/tooltips/ItemCompareFeature.java:
```diff
@@ -4,162 +4,590 @@
+        int slot = hoveredSlot.getContainerSlot();
+
+        // Selecting armor and accessory slots is not allowed
+        if (slot == 9 || slot == 10 || slot == 11 || slot == 12 || slot == 36 || slot == 37 || slot == 38 || slot == 39)
```
And long ifs should also have brackets.

kristofbolyai replied:
Also, I wonder, shouldn't InventoryModel keep track of these slots? I don't think it's great that they are hard-coded here, in such an "easy-to-miss" way.

kristofbolyai commented on common/src/main/java/com/wynntils/features/tooltips/ItemCompareFeature.java:
```diff
@@ -4,162 +4,590 @@
+        while (lines.getLast().getString().startsWith("§8")) {
+            lines.removeLast();
+        }
+    }
```
I'd like it so you refactor this method to use StyledText and check the first components actual ChatFormatting. I fear this could easily break if we changed how the lore component looks like upstream, `Component#getString` is not known to be the most stable result producing method..

uTa4u replied:
I'm having a bit of a problem doing this. In particular, I don't understand why lore lines have style= {color=dark_purple,!italic} if in `GearTooltipComponent#buildFooterTooltip` all lore `Components` have `ChatFormatting.DARK_GRAY` set as color

kristofbolyai replied:
Weird, I can take a look in a bit if you haven't figured it by then.

uTa4u replied:
I figured out that `GearTooltipComponent#buildFooterTooltip` is only used to build tooltips for fake items like ones that are displayed in Item Guide. So I guess lore having `dark_purple` as color in Style is some error(?) on Wynncraft's side

kristofbolyai commented on common/src/main/java/com/wynntils/features/tooltips/ItemCompareFeature.java:
```diff
@@ -4,162 +4,590 @@
+            StyledText line = StyledText.fromComponent(lines.get(i)).getNormalized();
+            if (line.getMatcher(WynnItemParser.SET_ITEM_PATTERN).matches()) {
+                lines.remove(i);
+            } else if (line.getMatcher(WynnItemParser.SET_PATTERN).matches()) {
+                lines.remove(i);
+                lines.remove(i - 1);
+                return;
```
Why remove just two lines? There are sets with more items in it. Or am I missing something here?

uTa4u replied:
> Why remove just two lines? There are sets with more items in it. Or am I missing something here?

First if branch removes item lines eg. "- Morph-Emerald" or "- Morph-Ruby". Second branch is when we hit set name eg. "Morph Set (1/8)", then we remove the set name and an empty line above it.

kristofbolyai reviewed (commented):
This code is actually really great as a first contribution! Thank you for working on it. I have some questions and overall code-style related reviews, however the overall picture is looking really good.

uTa4u committed:
Fixed code formatting, removed hardcoded slot numbers, refactored `removeFlavourText`

uTa4u commented:
Also found out that TM is not the only place where selected items behave weirdly. Bank container is also affected (maybe more). Better solution would be to clear selectedItems list when TM updates and when Bank page changes. In general when container's contents change. Maybe will look into it more later

kristofbolyai commented:
> Also found out that TM is not the only place where selected items behave weirdly. Bank container is also affected (maybe more). Better solution would be to clear selectedItems list when TM updates and when Bank page changes. In general when container's contents change. Maybe will look into it more later

What if you just cleared the selected items on `ScreenCloseEvent`? Iirc TM and bank works by "opening" a new container on page switches, and I think using this other event may solve it. If it doesn't, what if we just don't clear the selected item on screen close? I've kind of deliberately left that out from the original implementation, so you can compare items cross containers. I know it's not the best solution though..

kristofbolyai commented on common/src/main/java/com/wynntils/features/tooltips/ItemCompareFeature.java:
```diff
@@ -4,162 +4,634 @@
+            // Since line has dark_purple as color we get Component with actual color value with this
+            Component comp = StyledText.fromComponent(line).getComponent();
+            // This Component is empty, the Component we need is stored as it's sibling
+            List<Component> siblings = comp.getSiblings();
```
Sorry, but I'd still like you to change this up a little bit. Basically, use StyledText to get the first "StyledTextPart", and then turn it into a component, and check that color. StyledText does not gurantee that the resulting components always looks like (empty -> allchildren), it just guarantees that the output's style matches visually/in meaning what was passed in.

magicus commented on common/src/main/java/com/wynntils/features/tooltips/ItemCompareFeature.java:
```diff
@@ -4,162 +4,634 @@
+                        Models.Item.getWynnItem(armor).ifPresent(wynnItem -> {
+                            if (armor != hoveredItemStack) {
+                                itemsToCompare.add(new Pair<>(wynnItem, armor));
+                                ++equippedCount;
```
A style issue: Please use postfix increment, unless required by the context to read the value (which, btw, is discouraged). This applies to all these  ++equippedCount lines.

uTa4u committed:
FIxes

Fixed removeFlavourText, use postfix increment, event method name changes, fixed ItemStacks being duplicated in selectedItems, changed type of selectedItems

Fixed selected items not being removed from selectedItems properly

uTa4u committed:
ci: spotless formatting

ShadowCat117 commented on common/src/main/java/com/wynntils/features/tooltips/ItemCompareFeature.java:
```diff
@@ -4,162 +4,648 @@
 
     @SubscribeEvent
-    public void onItemTooltipRenderEvent(ItemTooltipRenderEvent.Post event) {
+    public void onHotbarSlotRenderEvent(HotbarSlotRenderEvent.Pre event) {
```
Is there a reason the selection arc needs to be drawn on the hotbar? Afaik this is for the hotbar in your hud not in the inventory and you clear selected items on closing containers

uTa4u replied:
For consistency only tbh. Since hotbar is still visible while inventory is open

ShadowCat117 replied:
I suppose it can be kept, I just don't see any reason to display that it's been selected twice when the focus will be on the inventory

ShadowCat117 commented on common/src/main/java/com/wynntils/features/tooltips/ItemCompareFeature.java:
```diff
@@ -4,162 +4,648 @@
+
+        int slot = hoveredSlot.getContainerSlot();
+
+        // Selecting armor and accessory slots is not allowed
+        for (int i : InventoryAccessory.getSlots()) {
+            if (slot == i) return;
+        }
```
What's the reason for this? If I wanted to compare my necklace with one in a container this forces me to select the container one which for some could be a minor inconvenience

uTa4u replied:
If by 'my necklace' you mean the one you have equipped, it is being compared against already. To compare with a necklace in a container just hold compare button over it

uTa4u replied:
Since equipped items are added to `itemsToCompare` first and only then selected items are added to it (if not present already) equipped items being selected will never change anything, so I just disabled selecting them. And the reason why I check if selected item is present in `itemsToCompare` is because held item (weapon) can be selected and not allowing to select held item is kinda weird (you can select it and then take it in your hand anyways)

kristofbolyai reviewed (approved):
Sorry for the delay!
